### PR TITLE
Update ring buffer API to allow static buffers

### PR DESCRIPTION
**Pull request by Kfeavel**
Cherry-picked from #206

### CODE REVIEW

**Comment by micahswitzer** [comment]
Any reason we can't use the template-based approach that I was suggesting earlier? I think it's an elegant way to solve this problem.

Something like the following (not sure if the template syntax is 100% correct)
```c++
template <int T>
class RingBuffer<T> {
private:
    uint8_t _data[T];
public:
    RingBuffer<T>();
}
```

**Comment by Kfeavel** [comment]
> Any reason we can't use the template-based approach that I was suggesting earlier? I think it's an elegant way to solve this problem.
> 
> Something like the following (not sure if the template syntax is 100% correct)
> 
> ```c++
> template <int T>
> class RingBuffer<T> {
> private:
>     uint8_t _data[T];
> public:
>     RingBuffer<T>();
> }
> ```

Resolved in 318e5d9.
Edit: Added size to template in 4243d47

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp [hunk 1]
```diff
@@ -13,84 +13,129 @@
...
+    }
+    /**
+     * @brief Query whether the circular buffer is empty.
+     *
+     * @return true The buffer is empty
+     * @return false The buffer is not empty
+     */
+    bool IsEmpty()
+    {
+        return (this->length == 0);
```
Comment (micahswitzer) [inline sysroot/usr/include/kernel/lib/ring_buffer.hpp]: Remove extra parens

File: kernel/dev/serial/rs232.cpp [hunk 2]
```diff
@@ -125,33 +129,11 @@ void rs232_init(uint16_t com_id) {
...
-
 char rs232_get_char() {
     mutex_lock(&mutex_rs232);
     // Grab the last byte and convert to a char
     uint8_t data = 0;
-    if (read_buffer != NULL)
-    {
-        ring_buffer_dequeue(read_buffer, &data);
-    }
+    ring.Dequeue(&data);
```
Comment (micahswitzer) [inline kernel/dev/serial/rs232.cpp]: Do we need to check the return code here?

File: kernel/dev/serial/rs232.cpp [hunk 3]
```diff
@@ -12,17 +12,18 @@
...
+#include <lib/mutex.hpp>
+#include <lib/string.hpp>
+#include <lib/ring_buffer.hpp>
 
-static ring_buff_t* read_buffer = NULL;
-uint8_t rs_232_line_index;
-uint16_t rs_232_port_base;
-mutex_t mutex_rs232("rs232");
+static uint16_t rs_232_port_base;
+static RingBuffer<uint8_t, 1024> ring;
```
Comment (micahswitzer) [inline kernel/dev/serial/rs232.cpp]: Why use `uint8_t` if we're enqueuing and dequeuing into `char`s?

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp [hunk 4]
```diff
@@ -13,84 +13,129 @@
...
+        return status;
+    }
+    /**
+     * @brief Removes a bytes from the circular buffer and writes
+     * the value of the byte to the data pointer.
+     *
+     * @param buf Buffer to contain the data
+     * @return int Returns 0 on success and -1 on error.
+     */
+    int Dequeue(T* buf)
```
Comment (micahswitzer) [inline sysroot/usr/include/kernel/lib/ring_buffer.hpp:61]: Can we maybe get an option to dequeue by value?

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp [hunk 5]
```diff
@@ -13,84 +13,129 @@
...
+    }
+    /**
+     * @brief Query whether the circular buffer is full.
+     *
+     * @return true The buffer is full
+     * @return false The buffer is not full
+     */
+    bool IsFull()
+    {
+        return (this->length == S);
```
Comment (micahswitzer) [inline sysroot/usr/include/kernel/lib/ring_buffer.hpp]: same as above

File: kernel/dev/serial/rs232.cpp (hunk 3 above)
Comment (Kfeavel) [reply kernel/dev/serial/rs232.cpp]: I feel like `uint8_t` is better for raw data, but there's no real reason. I can switch it so we can avoid a cast.

File: kernel/dev/serial/rs232.cpp (hunk 2 above)
Comment (Kfeavel) [reply kernel/dev/serial/rs232.cpp]: The worst thing that happens is a null character gets returned, and I'm not sure what else we would do besides that, so I'm not sure it's really worth checking the return value. What would we do in event of failure besides return 0 / null character?

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp (hunk 1 above)
Comment (Kfeavel) [reply sysroot/usr/include/kernel/lib/ring_buffer.hpp]: Resolved in 04f2009.

File: kernel/dev/serial/rs232.cpp (hunk 3 above)
Comment (Kfeavel) [reply kernel/dev/serial/rs232.cpp]: Resolved in 04f2009.

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp (hunk 4 above)
Comment (Kfeavel) [reply sysroot/usr/include/kernel/lib/ring_buffer.hpp:61]: Resolved in 04f2009.

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp (hunk 5 above)
Comment (Kfeavel) [reply sysroot/usr/include/kernel/lib/ring_buffer.hpp]: Resolved in 04f2009.

File: sysroot/usr/include/kernel/lib/ring_buffer.hpp [hunk 6]
```diff
@@ -13,84 +13,141 @@
...
+     * @return int Buffer capacity in bytes
+     */
+    int Capacity()
+    {
+        return S;
+    }
+
+private:
+    T data[S];
     int head;
```
Comment (micahswitzer) [inline sysroot/usr/include/kernel/lib/ring_buffer.hpp]: I know this isn't part of these changes, but could these be `size_t`s instead?

Comment (Kfeavel) [reply sysroot/usr/include/kernel/lib/ring_buffer.hpp]: Resolved in f95c4f0.

**Review by micahswitzer — APPROVED** [review]

**Pull request merged**